modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+                        }
+                    }
+
+                    addedSomething = true;
+                }
+            }
+        }
+
+        boolean droppedSomething = false;
+        {
+            // rip anything out of the task state that doesn't match what's in the cluster state,
+            // that is, if there's no longer an entry for a database in the repository,
+            // then drop it from the task state, too
+            Set<String> databases = geoIpMeta.getDatabases()
+                .values()
+                .stream()
+                .map(c -> c.database().name() + ".mmdb")
+                .collect(Collectors.toSet());
+            EnterpriseGeoIpTaskState _state = state;
+            Collection<Map.Entry<String, Metadata>> metas = List.copyOf(_state.getDatabases().entrySet());
+            for (Map.Entry<String, Metadata> entry : metas) {
+                String name = entry.getKey();
+                Metadata meta = entry.getValue();
+                if (databases.contains(name) == false) {
+                    logger.debug("Dropping [{}], databases was {}", name, databases);
+                    _state = _state.remove(name);
+                    deleteOldChunks(name, meta.lastChunk() + 1);
+                    droppedSomething = true;
+                }
+            }
+            state = _state;
+            updateTaskState();
+        }
+
+        if (addedSomething == false && droppedSomething == false) {
+            RuntimeException accumulator = null;
+            for (Map.Entry<String, DatabaseConfigurationMetadata> entry : geoIpMeta.getDatabases().entrySet()) {
+                final String id = entry.getKey();
+                DatabaseConfiguration database = entry.getValue().database();
+
+                final String accountId = database.maxmind().accountId();
+                try (HttpClient.PasswordAuthenticationHolder holder = credentialsBuilder.apply(accountId)) {
+                    if (holder == null) {
+                        logger.warn("No credentials found to download database [{}], skipping download...", id);
+                    } else {
+                        processDatabase(holder.get(), id, database);
+                    }
+                } catch (Exception e) {
+                    accumulator = ExceptionsHelper.useOrSuppress(accumulator, ExceptionsHelper.convertToRuntime(e));
+                }
+            }
+            if (accumulator != null) {
+                throw accumulator;
+            }
+        }
+    }
+
+    void processDatabase(PasswordAuthentication auth, String id, DatabaseConfiguration database) throws IOException {
+        final String name = database.name();
+        logger.debug("Processing database [{}] for configuration [{}]", name, id);
+
+        final String sha256Url = downloadUrl(name, "tar.gz.sha256");
+        final String tgzUrl = downloadUrl(name, "tar.gz");
+
+        final Pattern checksumPattern = Pattern.compile("(\\w{64})\\s\\s(.*)");
```

Comment:
This can be a static final variable at the class level instead of compiling it every time I think